### PR TITLE
feature/SIE94&95-created-emailverification-and-anonymousUser-plus-minor-fixes

### PR DESCRIPTION
**Pull request by bleachalon**
In this PR:
1. Moved signIn from userService to api/user.js. Added corresponding error handling to incorrect user credentials.
2. Added email verification when signUp and check authentication.
3. Added functionality to create the anonymous user and corresponding user doc.
4. Added functionality to link users who created an account to their anonymous user data.
5. Added signOut user functionality.
6. More detail of the above changes will be added to firebase Doc.

### CODE REVIEW

File: src/firebase/api/users.js [hunk 1]
```diff
@@ -173,3 +177,113 @@ export const updateUserData = async (updatedData) => {
...
+/**
+ * Sign current logged in user out.
+ * @return {Object} response object.
+ */
+export const signOut = async () => {
+  try {
+    await auth.signOut();
+    return {
+      status: StatusCodes.OK,
+      data: userAuth,
```
Comment (quanganhhoang) [inline src/firebase/api/users.js:222]: `data` should be null here? There is no `userAuth`

File: src/firebase/api/users.js [hunk 2]
```diff
@@ -173,3 +177,113 @@ export const updateUserData = async (updatedData) => {
...
+    return {
+      status: StatusCodes.INTERNAL_SERVER_ERROR,
+      data: null,
+      error,
+    };
+  }
+};
+
+/**
+ * This function is for users who do not what to sign in.
```
Comment (quanganhhoang) [inline src/firebase/api/users.js:235]: typo? Can you add more description of when and how to call this function? This feature is super cool though!

File: src/firebase/api/users.js [hunk 3]
```diff
@@ -173,3 +177,113 @@ export const updateUserData = async (updatedData) => {
...
+  }
+};
+
+/**
+ * Link user to their current anonymousAccount.
+ * @param {String} email email of user.
+ * @param {String} password password of user.
+ * @return {Object} userAuth or error object.
+ */
+export const linkUsetToAnonymousAccount = async (email, password) => {
```
Comment (quanganhhoang) [inline src/firebase/api/users.js:264]: typo

**Review by quanganhhoang — CHANGES_REQUESTED** [review]
a few comments but LGTM! Are these tested btw?

File: src/firebase/api/users.js (hunk 1 above)
Comment (bleachalon) [reply src/firebase/api/users.js:222]: > a few comments but LGTM! Are these tested btw?

Yes, they are all tested on main scenarios, but I may miss some corner cases, which I will continue testing. I will fix the typo and errors afternoon.

File: src/firebase/api/users.js [hunk 4]
```diff
@@ -173,3 +177,113 @@ export const updateUserData = async (updatedData) => {
...
+  try {
+    const credential = auth.EmailAuthProvider.credential(email, password);
+    const userAuth = await auth.currentUser.linkWithCredential(credential);
+    return {
+      status: StatusCodes.OK,
+      data: userAuth,
+      error: 'Linking success',
```
Comment (theodore3131) [inline src/firebase/api/users.js:271]: Shall we use 'message' or 'error' for this message? If so we can unify the return format. {status, data, error}

Comment (bleachalon) [reply src/firebase/api/users.js:271]: let's add this as a ticket for the next sprint to unify all responses.

File: src/firebase/api/users.js (hunk 2 above)
Comment (bleachalon) [reply src/firebase/api/users.js]: Changed it to be more meaningful and will add an example of using this function to the doc.

**Review by theodore3131 — APPROVED** [review]

**Comment by bleachalon** [comment]
> a few comments but LGTM! Are these tested btw?

**Pull request merged**